src/shogun/machine/GLM.h

```diff
@@ -0,0 +1,271 @@
+
+		bool m_compute_bias = true;
+
+		std::shared_ptr<GradientDescendUpdater> m_gradient_updater;
+
+		std::shared_ptr<ConstLearningRate> m_learning_rate;
+
+		std::shared_ptr<ElasticNetPenalty> m_penalty;
+
+		std::shared_ptr<GLMCostFunction> m_cost_function;
+};
+
+class GLMCostFunction
+{
+public:
+
+	friend class GLM;
+
+	GLMCostFunction()
+	{
+
+	}
+
+	virtual float64_t get_cost()
+	{
+		//TODO
+		return 0.0;
+	}
+
+	virtual SGVector<float64_t> get_gradient(const SGMatrix<float64_t> X, const SGVector<float64_t> y,
+												const SGVector<float64_t> w, const float64_t bias,
+												float64_t lambda, float64_t alpha, bool compute_bias,
+												float64_t eta, GLM_DISTRIBUTION distribution)
+	{
+		auto n_samples = y.vlen;
+	
+		auto z = compute_z(X, w, bias);
+		auto mu = non_linearity(z, compute_bias, eta, distribution);
+		auto grad_mu = gradient_non_linearity(z, eta, distribution);
+
+		SGVector<float64_t> grad_w(w.vlen);
+		SGVector<float64_t> a;
+		switch (distribution)
+		{
+		case POISSON:
+			{
+			a = linalg::element_prod(y, grad_mu);
+			for(auto i : range(y.vlen))
+				a[i] /= mu[i];
+
+			auto prod1 = linalg::matrix_prod(SGMatrix(grad_mu), X, true, true);
+			auto prod2 = linalg::matrix_prod(SGMatrix(a), X, true, true);
+
+			grad_w = linalg::transpose_matrix(linalg::add(prod1, prod2, 1.0, -1.0));
+			
+			break;
+			}
+		default:
+			error("Distribution type {} not implemented.", distribution);
+			break;
+		}
+
+		grad_w = linalg::scale(grad_w, 1.0/n_samples);
+
+		if(compute_bias)
+			grad_w = linalg::add(grad_w, w, 1.0, lambda * (1 - alpha));
+		else
+			grad_w = linalg::add(grad_w, w, 1.0, lambda * (1 - alpha));
+
+		return grad_w;
+	}
+
+	virtual float64_t get_gradient_bias(const SGMatrix<float64_t> X, const SGVector<float64_t> y,
+										const SGVector<float64_t> w, const float64_t bias,
+										bool compute_bias, float64_t eta, GLM_DISTRIBUTION distribution)
+	{
+		auto n_samples = y.vlen;
+		auto z = compute_z(X, w, bias);
+		auto mu = non_linearity(z, compute_bias, eta, distribution);
+		auto grad_mu = gradient_non_linearity(z, eta, distribution);
+
+		float64_t grad_bias = 0;
+		switch (distribution)
+		{
+		case POISSON:
+			for (auto i : range(grad_mu.vlen))
+			{
+				grad_bias += grad_mu[i];
+				grad_bias -= y[i] * grad_mu[i] / mu[i];
+			}
+			break;
+	
+		default:
+			error("Distribution type {} not implemented.", distribution);
+			break;
+		}
+		grad_bias /= n_samples;
+
+		return grad_bias;
+	}
+
+	virtual const char* get_name() const { return "GLMCostFunction"; }
+
+private:
+
+	virtual const SGVector<float64_t> compute_z(const SGMatrix<float64_t> X, const SGVector<float64_t> w,
+												const float64_t bias)
+	{
+		auto prod = linalg::matrix_prod(X, w, true);
+		linalg::add_scalar(prod, bias);
+		return prod;
+	}
+
+	virtual const SGVector<float64_t> non_linearity(const SGVector<float64_t> z, bool compute_bias,
+													float64_t eta, GLM_DISTRIBUTION distribution)
```

Comment:
you should do a round of `clang-format` to fix all these indentation/spacing issues

Reply:
Yes, I'll do this.